src/collective/collectionfilter/tiles/__init__.py

```diff
@@ -36,3 +48,111 @@ def filter_id(self):
     @property
     def reload_url(self):
         return self.url
+
+
+@implementer(ICollectionish)
+@adapter(ILayoutBehaviorAdaptable)
+class CollectionishLayout(CollectionishCollection):
+    """Provide interface for either objects with contentlisting tiles or collections or both"""
+
+    tile = None
+
+    def __init__(self, context):
+        """ Adapt either collections or contentlisting tile. The name is sorted content selector """
+        self.context = context
+
+        self.selectContent("")  # get first tile
+
+        if self.tile is None:
+            # Could still be a ILayoutAware collection
+            try:
+                self.collection = ICollection(self.context)
+            except TypeError:
+                raise TypeError("No contentlisting tile or Collection found")
```

Comment:
The `TypeError` is always raised and you cannot add a filtertile on a  `Page` context

Reply:
Fixed. not sure if there is a better way to tell queryadapter it couldn't adapt.

Reply:
erm ... there's a missunterstanding ... `zope.component._api.queryAdapter` has a kw parameter `default=None` so you do not get an Exception when using queryAdapter, simply it returns None if no one found ... you should remove this `raise` in line 71 because when you're editing a Page the first time, you do not have a chance to check if there's a contentlisting tile there ...

Reply:
Apparently you do get an exception with queryAdapter if one is raised in __init__ of the adapter. None tells you that no adapter was found, not that there was some problem in constructing the adapter and currently I can't see a way to indicate that an ILayoutAware doesn't have contentlisting other than raising an exception.

Reply:
ok ... but anyway this doesn't work when you initialy build a mosaic page with filter tiles ... you are not able to add filter tiles until you saved your page first with a contentlisting tile ... 

that's why I'd completely remove the check for existing contentlisting tiles while building/customizing a layout. When saving the layout I'd call an `IObjectModified` event which checks for mandatory contentlisting tile if a filtertile is in `ILayoutAware` adapter ... when it's missing -> status message.

Reply:
Thats what the change is supposed to do. I'll work out why its not working.

Reply:
ok. I've switched it to not throw exception anymore and use different way to determine if there is a listing available.
And added a test for it.